### PR TITLE
DEV-413 special DFX voting

### PR DESCRIPTION
**Pull request by ThaBrad**
Static Voting item for the special DFX voting

**--> end-date probably has to change** 

currently = `endDate: "2022-07-30T23:59:59.000Z"`

Will not be listed, at CFP page, after this date.

### CODE REVIEW

**Review by Krysh90 — APPROVED** [review]
Two short questions:
1. Why is this special CFP voting hard-coded in the app? Wouldn't it make more sense, to retrieve this information via API?
2. If it is hardcoded, what about translations?

**Comment by ThaBrad** [comment]
@Krysh90 
1. Because API simply fetches the stuff from GitHub and "forwards" it. No Persistence.
2. Only got the info now for that => added

**Comment by ThaBrad** [comment]
@davidleomay 
I still have no info about start and end date => currently hard-coded to
```
    startDate: "2022-07-23T23:59:59.000Z",
    endDate: "2022-07-30T23:59:59.000Z"
```

**Comment by ThaBrad** [comment]
Btw, I changed the staking balance check (and also updated the userDetailsDto).

From:
`setCanVote(stakingRoutes.find((r) => r.balance >= 100) != null);
`To:
`setCanVote(userDetail.stakingBalance >= 100);`

File: app/screens/SpecialVotingScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+>50% - DFX votet mit allen Masternodes „ja“
+Die Kollektivbetrachtung wird nur angewendet, so lange DFX <10 % der DeFiChain Masternodes für seine Kunden betreibt. Wird dieser Schwellenwert überschritten wird DFX automatisch die anteilige Betrachtung wählen
+
+
+Anteilige Betrachtung:
+
+Abhängig der Wahlergebnisse der jeweiligen CFPs von der DFX Community stimmt DFX für seine Kunden anteilig der abgegeben Stimmen mit ja/nein/neutral ab
+
+
+`
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:266]: Unused variables

File: app/screens/SpecialVotingScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+import { CompactRow, CompactCell } from "../elements/Tables";
+import { H1, H3, T5 } from "../elements/Texts";
+import withSession from "../hocs/withSession";
+import useAuthGuard from "../hooks/useAuthGuard";
+import { CfpResult, ResultStatus, VotingType } from "../models/CfpResult";
+import { CfpVote, CfpVotes } from "../models/User";
+import { getSettings, getUserDetail, putCfpVotes } from "../services/ApiService";
+import { Session } from "../services/AuthService";
+import NotificationService from "../services/NotificationService";
+import AppStyles from "../styles/AppStyles";
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:23]: Lot's of unused imports

File: app/screens/SpecialVotingScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  //     labels: [],
+  //     legend: [
+  //       `${t("cfp.yes")} (${Math.round((result.totalVotes.yes / result.totalVotes.total) * 100)}%)`,
+  //       `${t("cfp.neutral")} (${Math.round((result.totalVotes.neutral / result.totalVotes.total) * 100)}%)`,
+  //       `${t("cfp.no")} (${Math.round((result.totalVotes.no / result.totalVotes.total) * 100)}%)`,
+  //     ],
+  //     data: [[], [result.totalVotes.yes, result.totalVotes.neutral, result.totalVotes.no]],
+  //     barColors: [Colors.Success, Colors.LightGrey, Colors.Error],
+  //   };
+  // };
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:105]: Remove

File: app/screens/SpecialVotingScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    backgroundColor: Colors.Blue,
+    backgroundGradientFrom: Colors.Blue,
+    backgroundGradientTo: Colors.Blue,
+    decimalPlaces: 0,
+    color: (opacity = 1) => `rgba(255,  255,  255,  ${opacity})`,
+    labelColor: (opacity = 1) => `rgba(255,  255,  255,  ${opacity})`,
+    propsForBackgroundLines: {
+      stroke: Colors.Blue,
+    },
+  };
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:128]: Unused variable

File: app/screens/SpecialVotingScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        const currentDate = new Date()
+        const expiryDate = new Date(specialVotingCfp.endDate)
+
+        // specialVotingCfp automatically disappears after expiryDate
+        if (currentDate < expiryDate && currentDate > startDate) {
+          results.push(specialVotingCfp)
+        }
+
+        setCfpResults(results);
+        setVotes(votes);
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:86]: Useless line `setVotes(votes)`

File: app/screens/SpecialVotingScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+const DfxCfpNumbers: number[] = [];
+
+const SpecialVotingScreen = ({ session }: { session?: Session }) => {
+  const { t } = useTranslation();
+  const [isLoading, setIsLoading] = useState(true);
+  const [cfpFilter, setCfpFilter] = useState<"all" | "dfx">("all");
+  const [cfpResults, setCfpResults] = useState<CfpResult[]>();
+  const [canVote, setCanVote] = useState(false);
+  const [isVotingOpen, setIsVotingOpen] = useState(false);
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:33]: Unused state `isVotingOpen`

File: app/screens/SpecialVotingScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                  <RadioButton
+                    label={t("cfp.dfx_only")}
+                    onPress={() => setCfpFilter("dfx")}
+                    checked={cfpFilter === "dfx"}
+                  />
+                  <RadioButton label={t("cfp.all")} onPress={() => setCfpFilter("all")} checked={cfpFilter === "all"} />
+                </View>
+                <SpacerV height={50} />
+              </>
+            )}
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:152]: Unused code

File: app/screens/SpecialVotingScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+import withSession from "../hocs/withSession";
+import useAuthGuard from "../hooks/useAuthGuard";
+import { CfpResult, ResultStatus, VotingType } from "../models/CfpResult";
+import { CfpVote, CfpVotes } from "../models/User";
+import { getSettings, getUserDetail, putCfpVotes } from "../services/ApiService";
+import { Session } from "../services/AuthService";
+import NotificationService from "../services/NotificationService";
+import AppStyles from "../styles/AppStyles";
+
+const DfxCfpNumbers: number[] = [];
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:25]: Unused variable

File: app/screens/SpecialVotingScreen.tsx [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+const SpecialVotingScreen = ({ session }: { session?: Session }) => {
+  const { t } = useTranslation();
+  const [isLoading, setIsLoading] = useState(true);
+  const [cfpFilter, setCfpFilter] = useState<"all" | "dfx">("all");
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:30]: Unused state `cfpFilter`

File: app/screens/SpecialVotingScreen.tsx [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            {cfpResults
+              ?.filter((r) => cfpFilter === "all" || DfxCfpNumbers.includes(r.number))
+              ?.sort((a, b) => a.number - b.number)
+              .map((result) => (
+                <View key={result.number} style={{ width: "100%" }}>
+                  <View style={AppStyles.halfWidth} >
+                    <T5 text={t("specialVoting.explanation") + "\n\n\n"} /* style={AppStyles.center}  */ />
+                  </View>
+
+                  {(canVote || isSpecialVoting(result)) && (
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:170]: This is `SpecialVotingScreen`, `isSpecialVoting` is always true.

File: app/screens/SpecialVotingScreen.tsx [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+              </>
+            )}
+
+            {cfpResults
+              ?.filter((r) => cfpFilter === "all" || DfxCfpNumbers.includes(r.number))
+              ?.sort((a, b) => a.number - b.number)
+              .map((result) => (
+                <View key={result.number} style={{ width: "100%" }}>
+                  <View style={AppStyles.halfWidth} >
+                    <T5 text={t("specialVoting.explanation") + "\n\n\n"} /* style={AppStyles.center}  */ />
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:167]: Remove commented code

File: app/screens/SpecialVotingScreen.tsx [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+                          checked={(votes?.[result.number] ?? CfpVote.NEUTRAL) === CfpVote.NEUTRAL}
+                          disabled={!canVote}
+                          loading={isSaving?.number === result.number && isSaving.vote === CfpVote.NEUTRAL}
+                        />
+                      </View>
+                    </>
+                  )}
+
+                  <View style={AppStyles.halfWidth} >
+                    <T5 text={"\n\n\n" + t("specialVoting.description")} /* style={AppStyles.center}  */ />
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:201]: Remove commented code

File: app/screens/SpecialVotingScreen.tsx [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+  radioRow: {
+    flexDirection: "row",
+    alignItems: "center",
+    justifyContent: "space-between",
+    paddingHorizontal: 16,
+  },
+  voteContainer: {
+    justifyContent: "center",
+  },
+});
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:236]: Unused styles

File: app/i18n/pt.json [hunk 14]
```diff
@@ -339,6 +348,15 @@
...
       }
+    },
+    "specialVoting": {
+      "title": "",
+      "ended": "",
+      "link": "",
+      "collectively": "",
+      "proportionally": "",
+      "explanation": "",
+      "description": ""
```
Comment (Krysh90) [inline app/i18n/pt.json:359]: Empty values for pt?

File: app/components/HeaderContent.tsx [hunk 15]
```diff
@@ -58,6 +58,9 @@ const HeaderContent = ({ session, settings }: { session?: Session; settings?: Ap
           <DeFiButton onPress={() => navigate(Routes.Cfp)} style={styles.button} compact>
             {t("cfp.link")}
           </DeFiButton>
+          <DeFiButton onPress={() => navigate(Routes.SpecialVoting)} style={styles.button} compact>
+            {t("specialVoting.link")}
+          </DeFiButton>
```
Comment (Krysh90) [inline app/components/HeaderContent.tsx:63]: Just curious - will this button get removed again or at least hidden?  Or do we want to always show special votings even if there are no special votings active?

Comment (ThaBrad) [reply app/components/HeaderContent.tsx:63]: We'll remove it manually after a certain while. Currently we'll continue to show it to be transparent. 
We might use it thogh for future internal votings.

File: app/i18n/pt.json (hunk 14 above)
Comment (ThaBrad) [reply app/i18n/pt.json:359]: uff… I need to make that pre-comit script
But in any case, pt is not an officially supported language.

File: app/screens/SpecialVotingScreen.tsx [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+                <H3 text={t("specialVoting.ended")} />
+                <SpacerV height={50} />
+              </>
+            )}
+
+            {cfpResults
+              ?.sort((a, b) => a.number - b.number)
+              .map((result) => (
+                <View key={result.number} style={{ width: "100%" }}>
+                  <View style={AppStyles.halfWidth} >
```
Comment (davidleomay) [inline app/screens/SpecialVotingScreen.tsx:113]: Half-width looks IMO a bit weird on smaller screens:
image
I would propose to go with 90%.

Comment (davidleomay) [reply app/screens/SpecialVotingScreen.tsx:113]: There's a bug related to the halfWidth style. The background at the bottom is not blue:
image

**Review by davidleomay — CHANGES_REQUESTED** [review]

File: app/i18n/pt.json (hunk 14 above)
Comment (davidleomay) [reply app/i18n/pt.json:359]: Please add the PT translations before merging.

**Review by davidleomay — COMMENTED** [review]
Oli reported typos in the german translation. Use the following text:

---

Abstimmung: DFX Masternode CFP/DFIP Bewertungslogik

Jeder Nutzer des DFI Stakingservice von DFX erhält eine gewichtete Stimme auf Basis der investierten DFI. Zur Wahl stehen zwei alternative Abstimmungsmechanismen. Auswahlmöglichkeiten:

- Kollektiv: DFX soll mit allen Masternodes als Kollektiv mit ja/nein/neutral abstimmen
- Anteilig: DFX soll mit allen Masternodes anteilig der abgegebenen Stimmen mit ja/nein/neutral abstimmen

*** voting here ***

Kollektivbetrachtung: 
Abhängig der Wahlergebnisse der jeweiligen CFPs/DFIPs von der DFX Community stimmt DFX mit allen Masternodes für seine Kunden ab. 
< 50% - DFX votet mit allen Masternodes „nein“
50% - DFX votet mit allen Masternodes „neutral“
>50% - DFX votet mit allen Masternodes „ja“
Die Kollektivbetrachtung wird nur angewendet, so lange DFX <10 % der DeFiChain Masternodes für seine Kunden betreibt. Wird dieser Schwellenwert überschritten wird DFX automatisch die anteilige Betrachtung wählen

Anteilige Betrachtung:
Abhängig der Wahlergebnisse der jeweiligen CFPs von der DFX Community stimmt DFX für seine Kunden anteilig der abgegeben Stimmen mit ja/nein/neutral ab

**Review by davidleomay — COMMENTED** [review]
image
Title is duplicated.

**Review by davidleomay — COMMENTED** [review]
English text below

---

Voting: DFX Masternode CFP/DFIP Voting Logic.

Each user of DFX's DFI Staking Service receives a weighted vote based on their DFI invested. There are two alternative voting mechanisms to choose from. Choices:

- Collective: DFX shall vote yes or no with all masternodes as a collective.
- Proportional: DFX shall vote yes or no with all masternodes proportionate to the votes cast.

Your vote
Collective
Proportional
No preference



Collective View: 
Depending on the voting results of the respective CFPs/DFIPs from the DFX community, DFX votes with all masternodes for its customers. 

< 50% - DFX votes "no" with all masternodes.
>= 50% - DFX votes "yes" with all masternodes

Collective consideration is only applied as long as DFX operates <10% of DeFiChain masternodes for its customers. If this threshold is exceeded, DFX will automatically select proportional consideration

Proportional Consideration:
Depending on the voting results of the respective CFPs/DFIPs from the DFX community, DFX votes yes/no/neutral for its customers proportionally to the votes cast

File: app/screens/SpecialVotingScreen.tsx [hunk 17]
```diff
@@ -110,39 +112,39 @@ const SpecialVotingScreen = ({ session }: { session?: Session }) => {
...
-                      checked={(votes?.[result.number] ?? CfpVote.NEUTRAL) === CfpVote.NEUTRAL}
-                      disabled={!canVote}
-                      loading={isSaving?.number === result.number && isSaving.vote === CfpVote.NEUTRAL}
-                    />
-                  </View>
-
-                  <View style={AppStyles.halfWidth} >
-                    <T5 text={"\n\n\n" + t("specialVoting.description")} />
-                  </View>
+                  <T5 text={t("specialVoting.explanation") + "\n\n\n"} />
```
Comment (Krysh90) [inline app/screens/SpecialVotingScreen.tsx:115]: just curious - why using 3x new lines instead of `<SpacerV height={x} />`?

Comment (ThaBrad) [reply app/screens/SpecialVotingScreen.tsx:115]: - No new View in ViewHierarchy to render/compute
- Wanted to keep the relative line height for the paragraph

**Review by davidleomay — APPROVED** [review]

**Pull request merged**